### ISSUE TITLE
In the sampling, should we subtract the max first, for numerical stability?

### ISSUE THREAD

**Issue by hughperkins** [karpathy/char-rnn#133]
In the sampling code, we have:

```
-- use sampling
prediction:div(opt.temperature) -- scale by temperature
local probs = torch.exp(prediction):squeeze()
probs:div(torch.sum(probs)) -- renormalize so probs sum to one
prev_char = torch.multinomial(probs:float(), 1):resize(1):float()
```
... which is a softmax.

But in the third line, we are normalizing by dividing by the sum of exps, and I remember that it can sometimes be beneficial to subtact the max first, for numerical stability?  For example in your convnetjs, https://github.com/karpathy/convnetjs/blob/master/src/convnet_layers_loss.js#L31-L51 , we have:

```
// compute max activation
var as = V.w;
var amax = V.w[0];
for(var i=1;i<this.out_depth;i++) {
  if(as[i] > amax) amax = as[i];
}

// compute exponentials (carefully to not blow up)
var es = global.zeros(this.out_depth);
var esum = 0.0;
for(var i=0;i<this.out_depth;i++) {
  var e = Math.exp(as[i] - amax);
  esum += e;
  es[i] = e;
}

// normalize and output to sum to one
for(var i=0;i<this.out_depth;i++) {
  es[i] /= esum;
  A.w[i] = es[i];
}
```
Should we be doing the same thing here?

**Comment by karpathy** [comment]
The last layer of the LSTM model is 

```lua
local logsoft = nn.LogSoftMax()(proj)
```

which performs the normalization inside (I assume in a safe way), so this wouldn't be issue in this downstream code.

**Comment by hughperkins** [comment]
Ok, fair enough.
